cmd/influx_inspect/verify_seriesfile/verify_seriesfile.go

```diff
@@ -0,0 +1,298 @@
+			sLogger.Info("validating segment")
+
+			segment := tsdb.NewSeriesSegment(segmentID, segmentPath)
+			if err := segment.Open(); err != nil {
+				sLogger.Error("opening segment", zap.Error(err))
+				return false, nil
+			}
+			defer segment.Close()
+
+			if offset, err := cmd.validateSegment(segment, ids); err != nil {
+				sLogger.Error("iterating over segment",
+					zap.Int64("offset", offset),
+					zap.Error(err))
+				return false, nil
+			}
+
+			segments = append(segments, segment)
+		}
+
+		// validate the index if it exists
+		indexPath := filepath.Join(partitionPath, "index")
+		_, err = os.Stat(indexPath)
+		if os.IsNotExist(err) {
+			continue
+		}
+		if err != nil {
+			return false, err
+		}
+
+		pLogger.Info("validating index")
+
+		index := tsdb.NewSeriesIndex(indexPath)
+		if err := index.Open(); err != nil {
+			pLogger.Error("opening index", zap.Error(err))
+			return false, nil
+		}
+		defer index.Close()
+
+		if err := index.Recover(segments); err != nil {
+			pLogger.Error("recovering index", zap.Error(err))
+			return false, nil
+		}
+
+		// we check all the ids in a consistent order to get the same errors if
+		// there is a problem
+		idsList := make([]uint64, 0, len(ids))
+		for id := range ids {
+			idsList = append(idsList, id)
+		}
+		sort.Slice(idsList, func(i, j int) bool {
+			return idsList[i] < idsList[j]
+		})
+
+		for _, id := range idsList {
+			idData := ids[id]
+			idLogger := pLogger.With(zap.Uint64("id", id))
+
+			expectedOffset, expectedID := idData.offset, id
+			if idData.deleted {
+				expectedOffset, expectedID = 0, 0
+			}
+
+			// check both that the offset is right and that we get the right
+			// id for the key
+
+			if gotOffset := index.FindOffsetByID(id); gotOffset != expectedOffset {
+				idLogger.Error("index inconsistency",
+					zap.Int64("got_offset", gotOffset),
+					zap.Int64("expected_offset", expectedOffset))
+				return false, nil
+			}
+
+			if gotID := index.FindIDBySeriesKey(segments, idData.key); gotID != expectedID {
+				idLogger.Error("index inconsistency",
+					zap.Uint64("got_id", gotID),
+					zap.Uint64("expected_id", expectedID))
+				return false, nil
+			}
+		}
+	}
+
+	logger.Info("validation passed")
+	return true, nil
+}
+
+// idData keeps track of data about a series ID.
+type idData struct {
+	offset  int64
+	key     []byte
+	deleted bool
+}
+
+// validateSegment checks that all of the entries in the segment are well formed. If there
+// is any error, the offset at which it happened is returned.
+func (cmd *Command) validateSegment(segment *tsdb.SeriesSegment, ids map[uint64]idData) (offset int64, err error) {
+	defer func() {
+		if rec := recover(); rec != nil {
+			err = fmt.Errorf("panic validating segment: %v", rec)
+		}
+	}()
+
+	data := segment.Data()
+
+	advance := func(n int64) error {
+		if int64(len(data)) < n {
+			return fmt.Errorf("unable to advance %d bytes: %d remaining", n, len(data))
+		}
+		offset += n
+		data = data[n:]
+		return nil
+	}
```

Comment:
I'm personally not a fan of inline functions. I find it harder to read the code. I would either move the function out to a helper or inline the code to where it's used.